### PR TITLE
Change course_groups to refer to course_user instead of user

### PR DESCRIPTION
**Pull request by weiqingtoh**
This PR changes the `course_groups` to reference to `course_user`, mostly to support functionality at the course level (eg. leaderboard). 

Generally speaking most of it was replacement of `user` with `course_user`, but I've had to wrestle with many small things which were rather painful. I'm putting this up for review first, but I'm thinking if there are any specs we can add on for edge cases introduced by this change. 

I've split it up into 2 for easier review and changes; after that I will squash into a single commit.

### CODE REVIEW

File: app/models/course_user.rb [hunk 1]
```diff
@@ -41,6 +41,11 @@ class CourseUser < ActiveRecord::Base
     include CourseUser::AchievementsConcern
   end
   has_one :invitation, class_name: Course::UserInvitation.name, dependent: :destroy, validate: true
+  has_many :course_group_users, class_name: Course::GroupUser.name,
```
Comment (allenwq) [inline app/models/course_user.rb]: `has_many :group_users`, remove the course namespace so it's consistent with other associations.

File: app/models/components/course/groups_ability_component.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ def allow_group_manager_manage_group
   end
 
   def course_group_manager_hash
-    { group_users: { user_id: user.id, role: Course::GroupUser.roles[:manager] } }
+    { group_users: { course_user_id: user.course_user_ids,
```
Comment (lowjoel) [inline app/models/components/course/groups_ability_component.rb]: you can just do `{ group_users: { course_users: { id: user.id } }...`

File: app/models/course/group.rb [hunk 3]
```diff
@@ -27,8 +28,8 @@ def set_defaults
   # @return [Boolean]
   def should_create_manager?
     course && creator &&
-      course.course_users.exists?(user: creator) &&
-      !group_users.exists?(user: creator)
+      course.course_users.exists?(user_id: creator.id) &&
```
Comment (lowjoel) [inline app/models/course/group.rb]: why is this change needed?

File: app/models/course/group.rb [hunk 4]
```diff
@@ -27,8 +28,8 @@ def set_defaults
...
+      !group_users.map(&:course_user).map(&:user_id).include?(creator.id)
```
Comment (lowjoel) [inline app/models/course/group.rb]: `group_users.any? { |group_user| group_user.user == creator }`?

File: app/models/course/group_user.rb [hunk 5]
```diff
@@ -4,9 +4,9 @@ class Course::GroupUser < ActiveRecord::Base
 
   enum role: { normal: 0, manager: 1 }
 
-  validate :user_and_group_in_same_course
+  validate :course_user_and_group_in_same_course
 
-  belongs_to :user, inverse_of: :course_group_users
+  belongs_to :course_user, inverse_of: :group_users
   belongs_to :course_group, class_name: Course::Group.name, inverse_of: :group_users
```
Comment (lowjoel) [inline app/models/course/group_user.rb]: this should just be `belongs_to :group` right? @allenwq can you remember why `:course_group` was used?

File: spec/models/course/group_ability_spec.rb [hunk 6]
```diff
@@ -7,10 +7,11 @@
   with_tenant(:instance) do
     subject { Ability.new(user) }
     let(:course) { create(:course) }
+    let(:user) { create(:user) }
     let!(:group) { create(:course_group, course: course) }
 
     context 'when the user is a Course Staff' do
-      let(:user) { create(:course_manager, :approved, course: course).user }
+      before { create(:course_manager, :approved, course: course, user: user) }
```
Comment (lowjoel) [inline spec/models/course/group_ability_spec.rb]: this should be a let!

File: app/models/course/group_user.rb (hunk 5 above)
Comment (allenwq) [inline app/models/course/group_user.rb]: I checked the conversation, You said "there is a has_many through relation on the User model. groups on the user model do not make sense."

Comment (lowjoel) [inline app/models/course/group_user.rb]: but now that we are doing this on the course_user model, that statement no longer holds?

Comment (weiqingtoh) [inline app/models/course/group_user.rb]: Agree. Will change this to `belongs_to :group`.

File: app/models/course/group.rb (hunk 3 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: This one is required cause we can't do `group_user.user` for now. 

There are two solutions: 
1) We add a `has_one :user, through: :course_user` in the `group_user` model.
2) We access `user` through course_user, maybe define a method?

Comment (lowjoel) [inline app/models/course/group.rb]: what does `course.course_users` have to do with `group_user`?

File: app/models/components/course/groups_ability_component.rb (hunk 2 above)
Comment (weiqingtoh) [inline app/models/components/course/groups_ability_component.rb]: I tried that, but I suspect there is some issue with the construction of the query - probably to do with `cancancan-squeel` gem >_< 
```
NoMethodError:
  undefined method `klass' for nil:NilClass
  Did you mean?  class
  # /Users/weiqingtoh/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/cancancan-squeel-0.1.3/lib/cancancan/squeel/expression_builder.rb:68:in `build_association_comparison_node'
```

File: app/models/course/group.rb (hunk 3 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: Okay got it. Actually no change was required, will change back >_< Sorry. The `group_user.user` was also meant for the next line.

File: app/models/course/group.rb (hunk 4 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: I'm just going to copy this here -> 

>We can't do group_user.user for now.
>
>There are two solutions: 
>1) We add a has_one :user, through: :course_user in the group_user model.
>2) We access user through course_user, maybe define a method?

File: app/models/course/group_user.rb (hunk 5 above)
Comment (allenwq) [inline app/models/course/group_user.rb]: Yes 👍

File: app/models/components/course/groups_ability_component.rb (hunk 2 above)
Comment (lowjoel) [inline app/models/components/course/groups_ability_component.rb]: that indicates that the course_users association isn't defined on the group user model?

File: app/models/course/group.rb (hunk 4 above)
Comment (lowjoel) [inline app/models/course/group.rb]: group_user.course_user.user_id.

Comment (weiqingtoh) [inline app/models/course/group.rb]: Hmm okay. was thinking if we need `user` accessible through `group_user` - if not, what you suggested is 👍

File: app/models/components/course/groups_ability_component.rb (hunk 2 above)
Comment (weiqingtoh) [inline app/models/components/course/groups_ability_component.rb]: I'm not too sure, I manually tested on `rails c` and `Course::GroupUser.first.course_user` did return a valid `course_user`. Will investigate this further.

Comment (weiqingtoh) [inline app/models/components/course/groups_ability_component.rb]: [For my own use]

I tried some light debugging and ended up around this code block -> https://github.com/lowjoel/cancancan-squeel/blob/master/lib/cancancan/squeel/squeel_adapter.rb#L56-L59

```ruby
  def self.matches_relation?(subject, name, value)
    relation = subject.public_send(name)
    klass = subject.class.reflect_on_association(name).klass

    relation.where do
      expression, = CanCanCan::Squeel::ExpressionBuilder.build(self, klass, :==, value)
      expression
    end.any?
  end
```

Got lost at the line `relation.where do` (and some other information):
```ruby
subject = # group object
name = :group_users
value = { course_user: { user_id: 1}, role: 1 }
relation = # the group_user object
klass = Course::GroupUser
```

Comment (lowjoel) [inline app/models/components/course/groups_ability_component.rb]: Sorry. It should be singular, because it's a has_one.

```ruby
  { group_users: { course_user: { user_id: user.id },
                   role: Course::GroupUser.roles[:manager] } }
```

Comment (weiqingtoh) [inline app/models/components/course/groups_ability_component.rb]: Yup, apologies, the has I used was singular. The problem is that the query didn't reference the other tables. Will look into this more after I fix the other PRs.

```ruby
ActiveRecord::StatementInvalid:
       PG::UndefinedTable: ERROR:  missing FROM-clause entry for table "course_user"
       LINE 1: ..." WHERE "course_group_users"."group_id" = $1 AND ("course_us...
                                                                    ^
       : SELECT COUNT(*) FROM "course_group_users" WHERE "course_group_users"."group_id" = $1 AND ("course_user"."user_id" = 55 AND "course_group_users"."role" = 1)
     # ./spec/models/course/group_ability_spec.rb:29:in `block (4 levels) in <top (required)>'
     # ./spec/support/bullet.rb:7:in `block in extended'
     # ------------------
     # --- Caused by: ---
     # PG::UndefinedTable:
     #   ERROR:  missing FROM-clause entry for table "course_user"
     #   LINE 1: ..." WHERE "course_group_users"."group_id" = $1 AND ("course_us...
     #                                                                ^
     #   ./spec/models/course/group_ability_spec.rb:29:in `block (4 levels) in <top (required)>'
```

Comment (lowjoel) [inline app/models/components/course/groups_ability_component.rb]: weird. If you can't solve it I'll look at it on Thursday after my first paper.

**Comment by weiqingtoh** [comment]
Will push the correct version first; will have to wait for the PR to update squeel-cancancan adapter before we can re-review and merge this.

**Comment by lowjoel** [comment]
@weiqingtoh you can bundle update squeel-cancancan and commit that change with this PR.

**Comment by lowjoel** [comment]
@weiqingtoh sorry, I didn't see this. please rebase.

**Comment by weiqingtoh** [comment]
No worries. Rebased.

File: app/models/course/group.rb [hunk 7]
```diff
@@ -28,7 +29,7 @@ def set_defaults
   def should_create_manager?
     course && creator &&
       course.course_users.exists?(user: creator) &&
-      !group_users.exists?(user: creator)
+      !group_users.any? { |group_user| group_user.course_user.user_id == creator }
```
Comment (lowjoel) [inline app/models/course/group.rb]: this might be an N+1 query, might want to keep an eye out.

Comment (lowjoel) [inline app/models/course/group.rb]: as in, no need to fix now, just saying.

File: app/models/course/group_user.rb [hunk 8]
```diff
@@ -4,12 +4,12 @@ class Course::GroupUser < ActiveRecord::Base
...
-  validate :user_and_group_in_same_course
+  validate :course_user_and_group_in_same_course
 
-  belongs_to :user, inverse_of: :course_group_users
-  belongs_to :course_group, class_name: Course::Group.name, inverse_of: :group_users
+  belongs_to :course_user, inverse_of: :group_users
+  belongs_to :group, class_name: Course::Group.name, inverse_of: :group_users
 
-  alias_method :group, :course_group
+  # alias_method :group, :course_group
```
Comment (lowjoel) [inline app/models/course/group_user.rb]: @weiqingtoh remove this?

**Comment by lowjoel** [comment]
also, you can also probably squash it into one commit.

File: app/models/course/group.rb (hunk 7 above)
Comment (weiqingtoh) [inline app/models/course/group.rb]: Would the solution be `!group_users.course_users.includes(:user).any? { |group_user| group_user.course_user.user_id == creator }`?

Comment (lowjoel) [inline app/models/course/group.rb]: not sure, because the callback runs on creation of the record, and the group_user association may be partially loaded.
